### ISSUE TITLE
Regularization via constraint on weights instead (or in addition to) penalty term. 

### ISSUE THREAD

**Issue by mbkennel** [deeplearning4j/deeplearning4j#2505]
Request for enhancement or information for me to implement such an enhancement:

I would like the ability to apply a regularization constraint on weights.  This means that after taking a learning step in weight space, the weights are constrained to not exceed a certain L2 or L1 norm.  I

It is called 'max norm constraint' here: http://www.kdnuggets.com/2015/04/preventing-overfitting-neural-networks.html/2

I would like the values of the constraint parameter as supplied by the user to be on a 'per-weight' scale, and not a total sum scale.  So if one says the L2 constraint is 0.5, it means that 1/N \sum W_ij^2 <= 0.5 (or 0.5^2) with N the number of weights in W_ij == width * height.   If an L1 constraint it means  1/N \sum |W_ij|^2 <= 0.5 (or whatever the parameter is).   

If the constraint is violated (in the too big direction), then there is a truncation operation applied so move the weights as a whole back to the constraint surface, typically by the minimal L2 distance necessary. 

The constraint to L2 norm is simple, projection to a sphere, but constraint to L1 norm is not as obvious.  I know how to do it, but it requires an element-wise sorting procedure.  The question is where in the DL4J architecture this would go.  It cannot be computed on the raw gradient alone as it depends on the weight values after the gradient is applied. 

I would like this constraint to be able to be turned on and off (and its magnitude changed) layer by layer. I would like the constraint not to apply to any offset/bias term. 

The motivation for this enhancement is that typically "good" values of regularization constants are easier to find and tune in this formulation which tend to work well across multiple data sets.   With a penalty-based regularization one needs to balance the value of the error(cost) with the regularization penalty, but the quantitative magnitude of the error is often quite unknown when the modeler has started, and optimal values of regularization may be different to different orders of magnitude.  To the contrary, the typical sizes of weights in a "good" nnet (from prior experience) is often within a range of O(.1) to O(1) across a variety input datasets. 

A secondary consideration is that when starting training from small weights, the regularization is effectively off until training has proceeded further; this may speed training somewhat.

**Comment by AlexDBlack** [comment]
http://cs231n.github.io/neural-networks-2/#reg
"Max norm constraints. Another form of regularization is to enforce an absolute upper bound on the magnitude of the weight vector for every neuron and use projected gradient descent to enforce the constraint."

Looks like there's 2 options here, for enforcing the constraint:
- Similar to L1/L2: just another gradient/update component
- Enforce the constraint *after* the weight update has completed

First option is probably easier to implement given the current design/functionality - it's just another step in LayerUpdater (i.e., it's an additive term like l1/l2). Basically, push the constraint into the update... so perhaps something like:
temp = (weights + preMaxNormConstraintUpdates)
tempWeights = applyMaxNorm(temp)
difference = weights - tempWeights
finalUpdate = preMaxNormConstraintUpdates + difference
Again, that could be done as a step in Layer
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/updater/LayerUpdater.java#L98
Obviously a bit of overhead there,  but not much.


Second option: that would require an extra step (new functionality) - somewhere - to apply  that.
Currently the weights are modified only in one location: 
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/StochasticGradientDescent.java#L58
In other words, all parameter changes are currently done via the updates array.
So you'd look at the  weights after applying the update, and enforce the constraint by doing an in-place modification on the weights array.
I'd have to double check how that will interact with the line search optimizers though. Not sure yet exactly where would be best for that.

There's  some extra network configuration options to add too, but that's not hard.

**Comment by mbkennel** [comment]
The first option seems very prone to error.  Is the ‘gradient’ now before multiplying with the learning rate?  Before or after momentum?   It seems tricky to make sure it would be applied correctly.

The second would need a new hook but it seems general purpose enough.  Hinton also suggests a norm based constraint different from the ones I’ve been using typically;  it constrains the sum squared of weights going in to any given hidden node, this constraint is per node (not per layer).  This tries to ensure that each hidden unit roughly gets the same size and ‘worth’ of activations, and not a few overloaded and others ineffective.

And the hook could also be useful to output/log any sorts of diagnostics on the weights (such as, in particular, how close are the layers to hitting the constraints).

Actually, looking at the code briefly (given that I’ve been at this for all of 2 days), it appears that a subclass of GradientStepFunction could do it.  ProjectedGradientStepFunction.  That would need some access to configuration parameters per layer etc, but presumably that could be saved from a constructor as it would be constant in a run.  There would have to be some change in the constructor methodology as the values would be different for layers, and that’s not apparently currently supported.

As far as line search:  I’ve never implemented this with a line search optimizer (just SGD with momentum), but presumably the logic would be the same for any other constrainted optimization---in the line step if it attempts to make a step outside the feasible region then it should only take the step as far as it is allowed to do.

BTW:   I assume that in the code linked below ‘params’ are weights, and not external scalar parameters like learning rate?

cheers
Matt

**Comment by AlexDBlack** [comment]
Wouldn't call it especially error prone, it'd work fine IMO. If it's implemented in postApply, it'd be the final step - so after applying everything else (including momentum, l1/l2, learning rate etc).
I *would* call it inelegant and a inefficient though.


Alternatively: (I think I like this better) we add an 'enforce constraints' step after the weight update - pass in a model (MultiLayerNetwork, ComputationGraph, or Layer), and then (depending on the type of constraint) recursively apply the constraints to all layers.
The key differences between a 'contraint' and what might be termed an 'update component' (i.e., the set of functionality we have now) is that:
- a constraint wouldn't alter the score (only the model parameters) - consequently can't be gradient checked
- is applied after parameter updating is complete (hence modifies weights only)
- Doesn't require gradient/update or activation information to enforce the constraint
That mechanism/interface would generalize to other types of constraints too: for example capping the maximum value of weights or biases individually. In practice, I'm not sure there's that many of these types of 'hard/non-differentiable' constraints in practice? (I'd love to hear of more if you know any) - just the per-layer and per neuron weight constraints come to mind.

As for where to hook that in: probably in BaseOptimizer and StochasticGradientDescent just after the setParams calls
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/StochasticGradientDescent.java#L62
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/BaseOptimizer.java#L217


Finally, there's always the option of making a smooth (differentiable) version of the max norm constraint that is accounted for in the score (like L2 and L1 are). That wouldn't guarantee the constraint is satisfied at every step. It's an interesting idea (gradient-checkable too, for correctness), but borders more on original research...

**Comment by mbkennel** [comment]
I agree that I like the enforce constraints as a distinct phase.  I think it would be inappropriate for me to make significant architectural changes myself, though I could work on/test the specific regularizer constraints instantiations. 

| Finally, there's always the option of making a smooth (differentiable) version of the max norm constraint that is accounted for in the score (like L2 and L1 are). That wouldn't guarantee the constraint is satisfied at every step. It's an interesting idea (gradient-checkable too, for correctness), but borders more on original research...

That's essentially going back to the penalty function approach of regularization with a regularization term starting at a non-zero starting position instead of a zero.  In practice, it's much harder to get right because of the lack of knowledge of how big to set the penalty term. The traditional optimization community does have some pretty fancy algorithms and a long-term history in this area but that's opening up a whole new R&D project which I don't think is necessary.   A simple version could be done with the post apply framework above---the enforcement could be strict or soft (move weights closer but not necessarily into the feasible region)  I would prefer that it doesn't go into the score, it's more like a second score, i.e. "achieve this good performance, with violation of constraints of size y".  

One thing which would be very desirable is that the enforce constraints phase can return/print some diagnostic statistics.  At a basic level, I would want to see how close the nets are to having the constraints at the edge, if I see that the net isn't at the constraint boundary, it means that the constraint is not having any effect (which is OK, but you want to know how big the weights actually got before resetting the constraint tighter if you want to experiment). .

**Comment by AlexDBlack** [comment]
Re: return/print some diagnostic statistics:
I'd probably add that to TrainingListener: https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/api/TrainingListener.java
Most likely in the form of preEnforceConstraints(Model) and postEnforceConstraints(Model) methods. Diagnostic information shouldn't be build into the network itself - but a specific listener implementation can calculate whatever information it likes.

"I think it would be inappropriate for me to make significant architectural changes myself, though I could work on/test the specific regularizer constraints instantiations."
How about this: I can block out some of the methods and functionality and open a PR, then perhaps you could start from that, and work on filling in a few methods with the specific implementation (and adding tests etc)?

**Comment by mbkennel** [comment]
Hi, Alex, I now may have some more opportunity to work on this. 

Suppose I were to apply these constraints "manually", i.e. within the training loop of my code,  after each  call to model.train(miniBatch),   I would then artificially modify the weights of the 'model.params' to satisfy the constraint.   That would get ta POC of the functionality, though it wouldn't be integrated into the DL4J.

The real question is whether that would violate some assumptions about stored state of some algorithms, e.g. Nesterov momentum or some of the other updaters more complicated than memoryless plain SGD.   I asked this in the gitter but didn't have a response.

Unfortunately it is not possible to test this empirically because the result of the training should be distinctly different using the constraint-based regularization than without, and if it performs poorly it is a problem with the updater or the constraint?   We really need to understand the principles & assumptions of the updater and figure out which are OK to use, and if there is a bad result it is a fundamental algorithmic/problem conseqequence (constrained error surface is not suitable for the udpater in question) vs a bug we introduced.

**Comment by stolsvik** [comment]
A quick comment (and a vote!) for this: This is brought up in the original dropout paper:

http://jmlr.org/papers/volume15/srivastava14a.old/srivastava14a.pdf

> "One particular form of regularization was found to be especially useful for dropout— constraining the norm of the incoming weight vector at each hidden unit to be upper bounded by a fixed constant c. In other words, if w represents the vector of weights incident on any hidden unit, the neural network was optimized under the constraint ||w||_2 ≤ c. This constraint was imposed during optimization by projecting w onto the surface of a ball of radius c, whenever w went out of it. **This is also called max-norm regularization since it implies that the maximum value that the norm of any weight can take is c.** The constant c is a tunable hyperparameter, which is determined using a validation set. Max-norm regularization has been previously used in the context of collaborative filtering (Srebro and Shraibman, 2005). It typically improves the performance of stochastic gradient descent training of deep neural nets, even when no dropout is used."

**Comment by maxpumperla** [comment]
see https://github.com/deeplearning4j/deeplearning4j/issues/3898 as well

**Comment by AlexDBlack** [comment]
Implemented here:
https://github.com/deeplearning4j/deeplearning4j/pull/3957
